### PR TITLE
Tutorials update part 2

### PR DESCRIPTION
**Pull request by Yuriyzabegaev**
## Proposed changes

This pull request updates the tutorials of PorePy. Credits to @mariusnevland, @IngridKJ and @IvarStefansson 

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tutorials/Readme.md [hunk 1]
```diff
@@ -1,24 +1,32 @@
...
+Below we provide a general suggestion of in which order to read the tutorials.
+It should however be noted that the appropriate order of reading may depend on the reader's background and ambitions.
+
+1. [Introduction](./introduction.ipynb) describes the overarching conceptual framework and its high-level implementation. It also lists some problems which may be solved using PorePy.
+2. [Grids](./grids.ipynb) describes the structure of individual grids and demonstrates construction of different types of grids. It also shows how to access and manipulate grid quantities.
+3. [Grid topology](./grid_topology.ipynb) covers how to access various topological properties of grids in PorePy. This includes the relation between cells, faces and nodes, as well as the direction of the face normal vectors.
+4. [Meshing of fractures](./meshing_of_fractures.ipynb) describes the construction of mixed-dimensional grids. These grids represent a fracture network and the surrounding porous medium.
```
Comment (IvarStefansson) [inline tutorials/Readme.md:19]: Should we rename to "Mixed-dimensional grids"?

File: tutorials/Readme.md [hunk 2]
```diff
@@ -1,24 +1,32 @@
...
+5. [Conventions](./conventions.ipynb) defines some conventions used in PorePy. Specifically this is related to signs, boundary conditions etc.
+6. [Single phase flow](./single_phase_flow.ipynb) is where we introduce PorePy model classes and show how to run a simulation. It also covers how to do minor adjustments in a model.
+7. [Poromechanics](./poromechanics.ipynb) covers the concept of setting up a multiphysics simulation by reusing single physics model classes.
```
Comment (IvarStefansson) [inline tutorials/Readme.md:22]:
```suggestion
7. [Poromechanics](./poromechanics.ipynb) covers the concept of setting up a multiphysics simulation by reusing single-physics model classes.
```

File: tutorials/Readme.md [hunk 3]
```diff
@@ -1,24 +1,32 @@
...
-8. [contact_mechanics_with_AD](./contact_mechanics_with_AD.ipynb) delves deeper into the model classes and their use of the ad framework.
-9. [exporter](./exporter.ipynb) documents how to export data from PorePy for external visualization (e.g. ParaView).
-10. [equation_definition](./equation_definition.ipynb) details the steps required to set up and solve a mixed-dimensional partial differential equation with emphasis on the PorePy AD framework.
-11. [conventions](./conventions.ipynb) defines some conventions used in PorePy related to signs, boundary conditions etc.
+8. [Exporter](./exporter.ipynb) documents how to export data from PorePy for external visualization (e.g. ParaView). Several examples of exporting is demonstrated.
+9. [Exporting in models](./exporting_models.ipynb) shows how one can export data for visualization in PorePy model-based simulations.
+10. [Diagnostics](./diagnostics.ipynb) provides a description of how to use the diagnostics mixin in PorePy. The mixin allows for visualizing properties of the discretized system of equations in PorePy.
```
Comment (IvarStefansson) [inline tutorials/Readme.md:28]:
```suggestion
10. [Diagnostics](./diagnostics.ipynb) provides a description of how to use the diagnostics tool in PorePy. The tool allows for visualizing properties of the discretized system of equations in PorePy.
```

File: tutorials/Readme.md [hunk 4]
```diff
@@ -1,24 +1,32 @@
...
+11. [Equations](./equations.ipynb) briefly covers some general basics about automatic differentiation (AD). It mainly covers how AD is extensively used in PorePy equations, which means that it is most useful for the users that want to define their own equations.
+12. [Mandels problem](./mandels_problem.ipynb) shows how to set up and run the Mandel's consolidation problem based on the Biot equations of poroelasticity.
```
Comment (IvarStefansson) [inline tutorials/Readme.md:30]:
```suggestion
12. [Mandel's problem](./mandels_problem.ipynb) shows how to set up and run the Mandel's consolidation problem based on the Biot equations of poroelasticity. 
```

**Review by IvarStefansson — COMMENTED** [review]
Review part 1 (non-notebook). Minor comments.

File: tests/regression/check_tutorials.py [hunk 5]
```diff
@@ -1,74 +1,58 @@
...
+
+TUTORIAL_FILENAMES = glob.glob("tutorials/*.ipynb")
+
+
+@pytest.mark.parametrize("tutorial_path", TUTORIAL_FILENAMES)
+def test_run_tutorials(tutorial_path: str):
+    """We run the tutorial and check that it didn't raise any error.
+    This assumes we run pytest from the porepy directory.
+
+    Notice: the current file is now called "check_tutorials.py". Its name doesn't start
```
[the remark this answers is not in the document]
Comment (keileg) [reply tests/regression/check_tutorials.py:15]: I see no reason not to.

File: tutorials/grid_topology.ipynb [hunk 6]
```diff
@@ -0,0 +1,412 @@
+{
```
Comment (IvarStefansson) [inline tutorials/grid_topology.ipynb:1]: <p>are shown nicely</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (IvarStefansson) [inline tutorials/grid_topology.ipynb:1]: <p>sixth (counting from 0) node</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (IvarStefansson) [inline tutorials/grid_topology.ipynb:1]: <p>are found by</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (IvarStefansson) [inline tutorials/grid_topology.ipynb:1]: <p>Fix '</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

File: tutorials/grids.ipynb [hunk 7]
```diff
@@ -0,0 +1,520 @@
+{
```
Comment (IvarStefansson) [inline tutorials/grids.ipynb:1]: <p>"The parameter <code>info</code> decides which extra information is shown and the value of <code>alpha</code> determines the transparency of the grid.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (IvarStefansson) [inline tutorials/grids.ipynb:1]: <p>Link for the meshing tutorial.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

File: tutorials/meshing_of_fractures.ipynb [hunk 8]
```diff
@@ -56,10 +56,11 @@
    ]
```
Comment (IvarStefansson) [inline tutorials/meshing_of_fractures.ipynb:56]: <p>Consider renaming this tutorial to e.g. mixed_dimensional_grids.ipynb or fracture_networks.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (IvarStefansson) [inline tutorials/meshing_of_fractures.ipynb:56]: <p>Please fix grammar of changed sentence.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (IvarStefansson) [inline tutorials/meshing_of_fractures.ipynb:56]: <p>I don't see this as the most interesting feature. I think we could remove the export from the tutorial.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: tutorials/poromechanics.ipynb [hunk 9]
```diff
@@ -0,0 +1,429 @@
+{
```
Comment (IvarStefansson) [inline tutorials/poromechanics.ipynb:1]: <p>single-physics</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (IvarStefansson) [inline tutorials/poromechanics.ipynb:1]: <p>Eq. 1 is supplemented by eq. 2, which is the thermo-poromechanical stress. Thus, the reference might be somewhat confusing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (IvarStefansson) [inline tutorials/poromechanics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    class BoundaryConditionsMomentumBalance:
</pre><p>I suggest to not change the BCs of the momentum balance. The reason is that they are differentiated wrt time and should strictly speaking be TimeDependent (ad sense). I think the tutorial will not suffer from having the default conditions.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (IvarStefansson) [inline tutorials/poromechanics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #23.                    values.append(np.ones(sd.num_cells) * 0.1)
</pre><p>Please scale by appropriate units.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: tutorials/single_phase_flow.ipynb [hunk 10]
```diff
@@ -0,0 +1,557 @@
+{
```
Comment (IvarStefansson) [inline tutorials/single_phase_flow.ipynb]: <p>Please fix "In this tutorial will solve ". </p><p>Please include (and optionally improve) the following addition if you find it useful. "In PorePy, all class types except solution strategies are treated as Mixins. The solution strategy also defines an implicit contract between the functions used to run a simulation (run_time_dependent_problem) and the rest of the model."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (IvarStefansson) [inline tutorials/single_phase_flow.ipynb]: <p>Remove "only"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (IvarStefansson) [inline tutorials/single_phase_flow.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #7.            size = 2 / self.units.m
</pre><p>Very good! Please consider using this scaling in geometry-defining models in other tutorials.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (IvarStefansson) [inline tutorials/single_phase_flow.ipynb]: <p>run_time_dependent_model also calls other model methods. Please update.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (IvarStefansson) [inline tutorials/single_phase_flow.ipynb]: <p>I think "final solution" is better than "end solution".</p><p>Here, I would only reference the latter tutorial.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

File: tests/regression/check_tutorials.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply tests/regression/check_tutorials.py:15]: @keileg you mentioned that you're planning to add a separate github action for it. That's why I haven't renamed it yet - not to trigger it twice.

File: tutorials/single_phase_flow.ipynb (hunk 10 above)
Comment (keileg) [reply tutorials/single_phase_flow.ipynb]: <p>A slight modification: In PorePy all classes relating to models except solution strategies ...'</p>

File: tutorials/grids.ipynb [hunk 11]
```diff
@@ -0,0 +1,522 @@
...
```
Comment (keileg) [inline tutorials/grids.ipynb:1]: <p>Suggested additional sentence: The representation of grid geometry and topology is the same for Cartesian and simplex grids. This allows for setting up simulations, defining discretization methods etc., in ways that are agnostic to the problem dimension.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/878/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

**Comment by keileg** [comment]
@Yuriyzabegaev @IngridKJ @mariusnevland I went through the main tutorials, focusing on the big picture rather than details. I have left a few minor comments, but the overall impression is that this is very good. Thanks for the effort!

File: tests/regression/check_tutorials.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (keileg) [reply tests/regression/check_tutorials.py:15]: I will take care of it.

[the remark this answers is not in the document]
Comment (keileg) [reply tests/regression/check_tutorials.py]: Should be pushed now, please have a look @Yuriyzabegaev

File: tutorials/Readme.md [hunk 12]
```diff
@@ -1,24 +1,32 @@
 # Tutorials
-This folder contains tutorials that show the basic workings of PorePy. To follow the tutorials, you can download the files (either by cloning the repository, or download the individual files), and run them as Jupyter Notebooks.
+This folder contains tutorials to demonstrate how PorePy can be utilized.
+To follow the tutorials you can run them as Jupyter Notebooks, which is done either by downloading the individual files or by cloning the repository.
```
[the remark this answers is not in the document]
Comment (mariusnevland) [reply tutorials/Readme.md]: I have now changed this sentence to instruct users to specifically clone the repository.

**Comment by keileg** [comment]
@Yuriyzabegaev: It seeems something is wrong with the check of tutorials. Could you have a look, and let me know if it is an easy fix or if more thinking is needed?

File: src/porepy/numerics/time_step_control.py [hunk 13]
```diff
@@ -666,7 +666,7 @@ def is_schedule_in_simulated_times(
         return schedule.size == np.sum(in1d)
 
     # I/O
-    def write_time_information(self, path: Optional[Path] = None) -> None:
+    def write_time_information(self, path: Optional[Path | str] = None) -> None:
```
Comment (IvarStefansson) [inline src/porepy/numerics/time_step_control.py:669]: @keileg, can you comment on this one? Are we trying to enforce Path variables?

File: tutorials/grids.ipynb [hunk 14]
```diff
@@ -324,14 +319,14 @@
...
   {
-   "attachments": {},
    "cell_type": "markdown",
    "metadata": {},
    "source": [
     "The matplotlib interface is most useful for quick visualization, e.g. during debugging, and other visualization tools are recommended to be used otherwise. \n",
     "\n",
     "For showing information about cell numbers, node numbers, face numbers and face normals, we can assign values to the parameters `info` and `alpha` in `pp.plot_grid()`.\n",
-    "It is `info` that decides which extra information is shown, but as the value of `alpha` determines the transparency of the grid this value must also be chosen.\n",
+    "The parameter `info` decides which extra information is shown and the value of `alpha` determines the transparency of the grid\n",
```
Comment (IvarStefansson) [inline tutorials/grids.ipynb:328]:
```suggestion
    "The parameter `info` decides which extra information is shown and the value of `alpha` determines the transparency of the grid.\n",
```

File: tutorials/meshing_of_fractures.ipynb
Comment (IvarStefansson) [inline tutorials/meshing_of_fractures.ipynb:1]: Next time, use git mv to maintain file history.

File: tutorials/poromechanics.ipynb [hunk 15]
```diff
@@ -59,29 +62,38 @@
...
+    "                # Selecting central cells\n",
+    "                cell_centers = sd.cell_centers\n",
+    "                indices = (\n",
+    "                    (cell_centers[0] > (0.3 / units.m))\n",
+    "                    & (cell_centers[0] < (0.7 / units.m))\n",
+    "                    & (cell_centers[1] > (0.3 / units.m))\n",
+    "                    & (cell_centers[1] < (0.7 / units.m))\n",
+    "                )\n",
+    "\n",
+    "                acceleration = -9.8 * units.m / units.s**2\n",
```
Comment (IvarStefansson) [inline tutorials/poromechanics.ipynb:91]:
```suggestion
    "                acceleration = self.solid.convert_units(-9.8, "m * s^-2")\n",
```
Note that your solution is wrong, as the scaling needs to be inverted (9.8/units.m*units.s**2).

**Review by IvarStefansson — COMMENTED** [review]
Nice work! Accept with Minor Revisions.

File: src/porepy/numerics/time_step_control.py (hunk 13 above)
Comment (Yuriyzabegaev) [reply src/porepy/numerics/time_step_control.py:669]: @IvarStefansson this change is motivated by this line:
https://github.com/pmgbergen/porepy/blob/b92d5f98b1f313eaa1b96463c267d16c2749acfa/tests/unit/test_time_step_control.py#L659

Comment (IvarStefansson) [reply src/porepy/numerics/time_step_control.py:669]: I see. I still want Eirik's opinion, having heard him mention this recently.

File: tutorials/poromechanics.ipynb [hunk 16]
```diff
@@ -88,7 +88,7 @@
     "                    & (cell_centers[1] < (0.7 / units.m))\n",
     "                )\n",
     "\n",
-    "                acceleration = -9.8 * units.m / units.s**2\n",
```
Comment (IvarStefansson) [inline tutorials/poromechanics.ipynb]: There might be a \n missing. Sorry!

File: src/porepy/numerics/time_step_control.py (hunk 13 above)
Comment (keileg) [reply src/porepy/numerics/time_step_control.py:669]: File name should be represented as Paths. I will take care of this one.

File: tutorials/poromechanics.ipynb (hunk 15 above)
Comment (IvarStefansson) [reply tutorials/poromechanics.ipynb:91]: What happened here?

Comment (Yuriyzabegaev) [reply tutorials/poromechanics.ipynb:91]: I accepted your change and then added \n so now it works.

**Review by IvarStefansson — APPROVED** [review]

**Pull request merged**